axelor-sale/src/main/java/com/axelor/apps/sale/service/SaleOrderRestServiceImpl.java

```diff
@@ -0,0 +1,102 @@
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.collections.CollectionUtils;
+
+public class SaleOrderRestServiceImpl implements SaleOrderRestService {
+  protected SaleOrderRepository saleOrderRepository;
+  protected AppSaleService appSaleService;
+  protected CompanyService companyService;
+  protected SaleOrderInitValueService saleOrderInitValueService;
+  protected SaleOrderOnChangeService saleOrderOnChangeService;
+  protected SaleOrderDomainService saleOrderDomainService;
+  protected PartnerRepository partnerRepository;
+  protected SaleOrderLineGeneratorService saleOrderLineCreateService;
+
+  @Inject
+  public SaleOrderRestServiceImpl(
+      SaleOrderRepository saleOrderRepository,
+      AppSaleService appSaleService,
+      CompanyService companyService,
+      SaleOrderInitValueService saleOrderInitValueService,
+      SaleOrderOnChangeService saleOrderOnChangeService,
+      SaleOrderDomainService saleOrderDomainService,
+      PartnerRepository partnerRepository,
+      SaleOrderLineGeneratorService saleOrderLineCreateService) {
+    this.saleOrderRepository = saleOrderRepository;
+    this.appSaleService = appSaleService;
+    this.companyService = companyService;
+    this.saleOrderInitValueService = saleOrderInitValueService;
+    this.saleOrderOnChangeService = saleOrderOnChangeService;
+    this.saleOrderDomainService = saleOrderDomainService;
+    this.partnerRepository = partnerRepository;
+    this.saleOrderLineCreateService = saleOrderLineCreateService;
+  }
+
+  @Transactional(rollbackOn = {Exception.class})
+  @Override
+  public SaleOrder fetchAndAddSaleOrderLines(
+      List<SaleOrderLinePostRequest> saleOrderLinePostRequestList, SaleOrder saleOrder)
+      throws AxelorException {
+    if (CollectionUtils.isEmpty(saleOrderLinePostRequestList)) {
+      return saleOrder;
+    }
+    List<SaleOrderLine> saleOrderLineList =
+        createSaleOrderLine(saleOrderLinePostRequestList, saleOrder);
+    if (CollectionUtils.isNotEmpty(saleOrderLineList)) {
+      for (SaleOrderLine saleOrderLine : saleOrderLineList)
+        saleOrder.addSaleOrderLineListItem(saleOrderLine);
+    }
+    saleOrderRepository.save(saleOrder);
+    return saleOrder;
+  }
+
+  protected List<SaleOrderLine> createSaleOrderLine(
+      List<SaleOrderLinePostRequest> saleOrderLinePostRequestList, SaleOrder saleOrder)
+      throws AxelorException {
+    List<SaleOrderLine> saleOrderLineList = new ArrayList<>();
+    for (SaleOrderLinePostRequest saleOrderLinePostRequest : saleOrderLinePostRequestList) {
+      Product product = saleOrderLinePostRequest.fetchProduct();
+      BigDecimal quantity = saleOrderLinePostRequest.getQuantity();
+      SaleOrderLine saleOrderLine =
+          saleOrderLineCreateService.createSaleOrderLine(saleOrder, product, quantity);
```

Comment:
Use SaleOrderLineGeneratorService instead. Then you won't need to add the line to the sale order.

The method should not return a list, juste create a SoL for every SaleOrderLinePostRequest.